qe/drivers.py

```diff
@@ -1707,3 +1866,682 @@ def selection_step(
 
     # 3.
     return (*Powerplant(lewis, psi_det_extented).E_and_psi_coef, psi_det_extented)
+
+
+#  ___  _________ _____
+#  |  \/  || ___ \_   _|
+#  | .  . || |_/ / | |
+#  | |\/| ||  __/  | |
+#  | |  | || |    _| |_
+#  \_|  |_/\_|    \___/
+#
+
+
+def dispatch_psi(comm, psi_i):
+    """Function to partition and dispatch pieces of trial wavefunction psi_i to worker processes
+
+    :param comm: mpi4py.MPI.COMM_WORLD communicator object
+    :param psi_i: Psi_det, list of determinants"""
+
+    rank = comm.Get_rank()  # Rank of current process
+    size = comm.Get_size()  # No. of processes currently running
+
+    if rank == 0:
+        indices = np.arange(len(psi_i), dtype="i")
+        number_of_dets, res = divmod(len(psi_i), size)
+        # Number of determinants sent to each worker; First `res` processes sent one extra
+        sendcounts = np.array(
+            [number_of_dets + 1 if i < res else number_of_dets for i in range(size)]
+        )
+        displacement = np.array(
+            [sum(sendcounts[:i]) for i in range(size)]
+        )  # Entry i of displacement vector is the index of the first determinant process i receives
+    else:
+        indices = None  # Worker processes send nothing
+        sendcounts = np.zeros(size, dtype="i")  # Preallocate space for count on workers
+        displacement = None
+
+    sendcounts = comm.bcast(
+        sendcounts, root=0
+    )  # Broadcast count to each process, which will need this to preallocate appropriate space
+    local_indices = np.zeros(
+        sendcounts[rank], dtype="i"
+    )  # Preallocate space for process to receive count[rank] determinants
+    comm.Scatterv([indices, sendcounts, displacement, MPI.INT], local_indices, root=0)
+
+    # Create new list of local determinants on each node
+    psi_local = []
+    for I in local_indices:
+        psi_local.append(psi_i[I])
+
+    return psi_local
+
+
+@dataclass
+class Hamiltonian_one_electron_manager(object):
+    """Instance of `Hamiltonian_one_electron()` class for particular
+    lists of internal/external determinants. Used to cache `local' Hamiltonian.
+    Re-created at each CIPSI iteration."""
+
+    psi_i: Psi_det  # Internal (or local) list of determinants
+    psi_j: Psi_det  # External (or full internal) list of determinants
+    d_one_e_integral: One_electron_integral
+    E0: Energy
+
+    def __init__(
+        self,
+        psi_i: Psi_det,
+        psi_j: Psi_det,
+        d_one_e_integral: One_electron_integral,
+        E0: Energy,
+        comm: MPI.COMM_WORLD = None,
+    ):
+        self.d_one_e_integral = d_one_e_integral
+        self.E0 = E0
+        if comm is None:  # Non-MPI jobs
+            self.comm = None
+            self.psi_i = psi_i
+            self.psi_j = psi_j
+        else:  # For MPI jobs
```

Comment:
I think now that we have only a mpi Davison, the code will break non the less, you can just remove this if. We need to have mpi to run the code